File: c/src/sorting/merge_sort.c
#include "algorithms.h"

int *merge(int *a, int p, int q, int r) {
  int n1 = q - p + 1;
  int n2 = r - q;

  int *left = malloc((n1 + 1) * sizeof(int));
  int i;
  for (i = 0; i < n1; i++) {
    left[i] = a[p + i];
  }
  left[i] = INT_MAX;

  int *right = malloc((n2 + 1) * sizeof(int));
  int j;
  for (j = 0; j < n2; j++) {
    right[j] = a[q + j + 1];
  }
  right[j] = INT_MAX;

  i = 0;
  j = 0;
  for (int k = p; k <= r; k++) {
    if (left[i] <= right[j]) {
      a[k] = left[i];
      i += 1;
    } else {
      a[k] = right[j];
      j += 1;
    }
  }

  free(left);
  free(right);

  return a;
}

int *merge_sort_rec(int *a, int p, int r) {
  if (p < r) {
    int q = (p + r) / 2;
    merge_sort_rec(a, p, q);
    merge_sort_rec(a, q + 1, r);
    return merge(a, p, q, r);
  }
  return a;
}

int *merge_sort(int *a, int size) { return merge_sort_rec(a, 0, size - 1); }

File: rust/src/sorting/merge_sort.rs
use std::i32;

pub fn sort(a: &mut Vec<i32>, size: usize) {
  return merge_sort(a, 0, size - 1);
}

fn merge_sort(a: &mut Vec<i32>, p: usize, r: usize) {
  if p < r {
    let q = (p + r) / 2;
    merge_sort(a, p, q);
    merge_sort(a, q + 1, r);
    return merge(a, p, q, r);
  }
}

fn merge(a: &mut Vec<i32>, p: usize, q: usize, r: usize) {
  let n1 = q - p + 1;
  let n2 = r - q;

  let mut left = vec![0; n1 + 1];
  for i in 0..n1 {
    left[i] = a[p + i as usize]
  }
  left[n1] = i32::MAX;

  let mut right = vec![0; n2 + 1];
  for j in 0..n2 {
    right[j] = a[q + j + 1 as usize]
  }
  right[n2] = i32::MAX;

  let mut i = 0;
  let mut j = 0;
  for k in p..=r {
    if left[i] <= right[j] {
      a[k] = left[i];
      i += 1
    } else {
      a[k] = right[j];
      j += 1
    }
  }
}
